### ISSUE TITLE
Issues with DoT

### ISSUE THREAD

**Issue by bortzmeyer** [BRE-ISNIC/bre-doh-analysis#3]
"DNS over TLS differs from DNS over HTTPS, mostly in that it omits the HTTP/2 protocol, sending DNS data directly over the TLS circuit itself. This simplification means DoT misses the opportunity to build upon the tooling and experience provided by the web community, while inheriting most of the same technical challenges (certificate management, implementation complexity, and performance penalties) from TCP and TLS." It seems to mention only DoT problems and limits. It could be turned around: by not using HTTP, DoH is leaner and simpler (and leaks less personal data).

**Comment by BRE-ISNIC** [comment]
You are correct, the document does not spend much effort on extolling the virtues of DoH's alternatives.

That is deliberate: My goal here is to try and explain why DoH is successful where DoT was not.

Being leaner and simpler didn't contribute to DoT winning (DoH is winning), so raising that point doesn't (in my opinion) facilitate understanding of how we got where we are. Do you disagree? It can be hard to estimate which arguments are important and which are not. I felt that one could be safely omitted, but I could be convinced otherwise.

**Comment by jedisct1** [comment]
One reason DoT is not successful is that many public WiFi hotspots only allow ports 80 and 443. So DoT cannot be used at all where it would actually be useful.

DoT doesn't encapsulate DNS packets. It doesn't even add a version header that could allow further revisions of the protocol. As a result, the inner protocol is stuck forever; it can never be improved.

Due to that design, parsing and rewriting queries and responses is required in order to handle padding. This is an ugly last-minute hack (padding is related to transport, it shouldn't have anything to do in a RR set), that undermines the apparent simplicity of the protocol. As a result, most implementations do not even implement padding at all.

Besides tooling and experience, HTTP/2 supports multiplexing and prioritization out of the box. Something impossible without reinventing these mechanisms in DoT. Which would not even be possible since DoT cannot be extended.

So, DoT was dead on arrival.

But more importantly, HTTP is the protocol that has the largest ecosystem, the best infrastructure (including caches), and is what all the optimizations efforts are focused on. Building upon it makes a lot of sense.

**Comment by bortzmeyer** [comment]
DoT has multiplexing, like good old DNS-over-TCP and for the same reason (that's why the Query ID exists, while it is always zero for DoH, which relies on HTTP-specific tricks).

**Comment by jedisct1** [comment]
It doesn't. Other data cannot be transported on the same connection. HTTP does, and that's one of the main selling point of DoH.

**Comment by bortzmeyer** [comment]
"Other", you mean not DNS? That's questionable, since it may give a lot of information to the resolver. (See Internet-Draft draft-dickinson-doh-dohpe)

**Comment by jedisct1** [comment]
The goal is to avoid extra round trips when the same CDN acts as a resolver and has the website content (see I-dont-know what draft, another-one-about-using-a-specific-resolver-for-a-set-of-domains and the one about sending certificates that way, too).

**Comment by jedisct1** [comment]
On giving a lot of information to resolvers, DoT is the worst candidate for anonymization. It cannot be served as an onion service, an anonymization extension doesn’t exist (and since the protocol cannot be extended...), and zero efforts have ever been made to anonymize it anyway. At best, it can rely on Oblivious DNS that has issues of its own.

OTOH, Oblivious DoH exists, and could be made because it was based on HTTP.

**Comment by bortzmeyer** [comment]
For a .onion service, running plain DNS-over-TCP seems the most reasonable thing to do, since Tor handles encryption, in that case (I know there is a debate about whether HTTP onion sites should use http or https but I won't discuss it here).

Otherwise, DoT works fine over Tor. (I just tested.)

**Comment by jedisct1** [comment]
Yes, using exit nodes, not as an onion service.

**Comment by bortzmeyer** [comment]
I disagree, I just set  up an onion service and it works (lani4a4fr33kqqjeiy3qubhfx2jewfd3aeaepuwzxrx6zywp2mo4cjad.onion) 

[Only problem, the real server is too impatient and tends to close the TCP connection too fast. Add +keepopen to dig or +nofastopen to kdig solves the issue.]
